docs/custom-components.mdx

```diff
@@ -0,0 +1,286 @@
+        };
+        Ok(output)
+    }
+}
+```
+
+Components must implement the `Guest` trait, which is the main interface between your component and the WAVS runtime. The `run` function is the entry point for processing triggers: it receives the trigger data, decodes it, processes it according to your component's logic, and returns the results. Results can be sent back to Ethereum (when triggered by a contract event) or returned as CLI output (when triggered manually during testing). Error handling is done through the `Result` type. In the foundry template, bindings are automatically generated for any files in the `/components` and `/src` directories when the `make build` command is run.
+
+### Building Components
+
+Dependencies for components are added to the `Cargo.toml` file in the `components` directory.
+
+```toml
+[dependencies]
+# Essential dependencies
+wit-bindgen-rt = {workspace = true}    # Required for WASI bindings and Guest trait
+wavs-wasi-chain = { workspace = true }  # Required for core WAVS functionality
+# Optional dependencies
+serde = { workspace = true }            # For serialization (if working with JSON)
+serde_json = { workspace = true }       # For JSON handling
+alloy-sol-macro = { workspace = true }  # For Ethereum contract interactions
+wstd = { workspace = true }             # For WASI standard library features
+alloy-sol-types = { workspace = true }  # For Ethereum ABI handling
+anyhow = { workspace = true }           # For enhanced error handling
+```
+
+#### Testing
+
+Components can be built for local testing using the following commands. The `wasi-exec` command allows you to simulate triggers, and you can view the component's logs in stdout/stderr to debug its execution.
+
+```bash
+# Build the component
+make build
+
+# Test locally
+make wasi-exec
+```
+
+### `wavs-wasi-chain` crate
+
+The `wavs-wasi-chain` crate provides a set of helpful functions for making HTTP requests and interacting with the blockchain. It also provides a macro for decoding trigger data for use in the component.
+
+Learn more in the [crate documentation](https://docs.rs/wavs-wasi-chain/latest/wavs_wasi_chain/all.html#functions).
+
+### Network requests
+
+Components can make network requests to external APIs using the `wavs-wasi-chain` crate.
+
+The following dependencies are useful for making HTTP requests from a component. These are added to a component's `Cargo.toml` file:
+
+```toml Cargo.toml
+[dependencies]
+wavs-wasi-chain = { workspace = true }  # HTTP utilities
+wstd = { workspace = true }             # Runtime utilities
+serde = { workspace = true }            # Serialization
+serde_json = { workspace = true }       # JSON handling
+```
+
+The following example shows how to make a basic HTTP GET request from a component:
+
+```rust lib.rs
+async fn make_request() -> Result<YourResponseType, String> {
+    // Create the request
+    let url = "https://api.example.com/endpoint";
```

Comment:
@dakom Does this example make sense?

Reply:
yep, it looks right to me!

That said, we have a helper for POST requests that expect JSON.. in the [`Permissions` component example we have](https://github.com/Lay3rLabs/WAVS/blob/162abc479d1474025c1aa01d1ff8d3dff04396e3/examples/components/permissions/src/lib.rs#L59):

```
let post_response: PostResponse =
  fetch_json(http_request_post_json(&input.post_url, &input.post_data)?).await?;
```

It's a bit unfortunate that we don't have `http_request_get_json` in that repo now, it would make the example here much simpler (manually adding `Accept: application/json` is a bit lame for people looking at our helpers).. I'll open an issue for that